crypto/random/chacha20.go

```diff
@@ -0,0 +1,194 @@
+package random
+
+import (
+	"encoding/binary"
+	"fmt"
+
+	"golang.org/x/crypto/chacha20"
+)
+
+// We use Chacha20, to build a cryptographically secure random number generator
+// that uses the ChaCha algorithm.
+//
+// ChaCha is a stream cipher designed by Daniel J. Bernstein[^1], that we use as a PRG. It is
+// an improved variant of the Salsa20 cipher family.
+//
+// We use Chacha20 with a 256-bit key, a 192-bit stream identifier and a 32-bit counter as
+// as specified in RFC 8439 [^2].
+// The encryption key is used as the PRG seed while the stream identifier is used as a nonce
+// to customize the PRG. The PRG outputs are the successive encryptions of a constant message.
+//
+// A 32-bit counter over 64-byte blocks allows 256 GiB of output before cycling,
+// and the stream identifier allows 2^192 unique streams of output per seed.
+// It is the caller's responsibility to avoid the PRG output cycling.
+//
+// [^1]: D. J. Bernstein, [*ChaCha, a variant of Salsa20*](
+//       https://cr.yp.to/chacha.html)
+//
+// [^2]: [RFC 8439: ChaCha20 and Poly1305 for IETF Protocols](
+//       https://datatracker.ietf.org/doc/html/rfc8439)
+
+// The PRG core, implements the randCore interface
+type chachaCore struct {
+	cipher chacha20.Cipher
+
+	// empty message added to minimize allocations and buffer clearing
+	emptyMessage [lenEmptyMessage]byte
+
+	// Only used for State/Restore functionality
+
+	// Counter of bytes encrypted so far by the sream cipher.
+	// Note this is different than the internal 32-bits counter of the chacha state
+	// that counts the encrypted blocks of 512 bits.
+	bytesCounter uint64
+	// initial seed
+	seed [keySize]byte
+	// initial customizer
+	customizer [nonceSize]byte
+}
+
+// The main PRG, implements the Rand interface
+type chachaPRG struct {
+	genericPRG
+	core *chachaCore
+}
+
+const (
+	keySize   = chacha20.KeySize
+	nonceSize = chacha20.NonceSize
+
+	// Chacha20SeedLen is the seed length of the Chacha based PRG, it is fixed to 32 bytes.
+	Chacha20SeedLen = keySize
+	// Chacha20CustomizerMaxLen is the maximum length of the nonce used as a PRG customizer, it is fixed to 24 bytes.
+	// Shorter customizers are padded by zeros to 24 bytes.
+	Chacha20CustomizerMaxLen = nonceSize
+)
+
+// NewChacha20PRG returns a new Chacha20-based PRG, seeded with
+// the input seed (32 bytes) and a customizer (up to 12 bytes).
+//
+// It is recommended to sample the seed uniformly at random.
+// The function errors if the seed is different than 32 bytes,
+// or if the customizer is larger than 12 bytes.
+// Shorter customizers than 12 bytes are padded by zero bytes.
+func NewChacha20PRG(seed []byte, customizer []byte) (*chachaPRG, error) {
+
+	// check the key size
+	if len(seed) != Chacha20SeedLen {
+		return nil, fmt.Errorf("chacha20 seed length should be %d, got %d", Chacha20SeedLen, len(seed))
+	}
```

Comment:
Can we restrict the type and eliminates the error case here?
```suggestion
func NewChacha20PRG(seed [Chacha20SeedLen]byte, customizer []byte) (*chachaPRG, error) {
```


or 
```
func NewChacha20PRG(seed [32]byte, customizer []byte) (*chachaPRG, error) {
```

Reply:
That is possible, though it means we will have the copy the seed slice into an array to pass it to `NewChacha20PRG` each time in the protocol code.
I'm planning to make changes to the protocol in a follow-up issue, so I added an item for this: https://github.com/dapperlabs/flow-go/issues/6107